api/service/dashboard/dashboard.go

```diff
@@ -0,0 +1,1155 @@
+package dashboard
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/dezswap/dezswap-api/pkg/db/aggregator"
+	"github.com/dezswap/dezswap-api/pkg/db/parser"
+	"github.com/dezswap/dezswap-api/pkg/dezswap"
+	"github.com/pkg/errors"
+	"gorm.io/gorm"
+)
+
+type dashboard struct {
+	chainId string
+	*gorm.DB
+}
+
+var chartCriteriaByDuration = map[Duration]struct {
+	Ago     string
+	TruncBy time.Duration
+}{
+	// a month range with everyday data
+	Month: {"1 month", time.Hour * 24},
+	// 3 months range with every week data
+	Quarter: {"3 month", time.Hour * 24 * 7},
+	// 1 year range with every 2 weeks data
+	Year: {"1 year", time.Hour * 24 * 7 * 2},
+	// 10 years range with every month data
+	All: {"10 year", time.Hour * 24 * 31},
+}
+
+var _ Dashboard = &dashboard{}
+
+func NewDashboardService(chainId string, db *gorm.DB) Dashboard {
+	return &dashboard{chainId, db}
+}
+
+// Aprs implements Dashboard.
+func (d *dashboard) Aprs(duration Duration) (Aprs, error) {
+	truncBy := int64(chartCriteriaByDuration[duration].TruncBy.Truncate(time.Second).Seconds())
+	intervalAgo := chartCriteriaByDuration[duration].Ago
+
+	dateSeries := fmt.Sprintf(`
+		SELECT
+			generate_series(
+				CAST(FLOOR(EXTRACT(EPOCH FROM DATE_TRUNC('day', NOW() ))) AS int8),
+				CAST(FLOOR(EXTRACT(EPOCH FROM DATE_TRUNC('day', NOW() - INTERVAL '%s'))) AS int8),
+			-%d
+			) AS timestamp
+		`, intervalAgo, truncBy)
+	query := fmt.Sprintf(`
+	WITH ds AS (%s),
+	tvl AS (
+		SELECT
+			SUM(liquidity) AS tvl,
+			t.timestamp
+		FROM (
+			SELECT DISTINCT ON (ps.pair_id, ds.timestamp)
+				ps.pair_id AS pair_id,
+				liquidity0_in_price + liquidity1_in_price AS liquidity,
+				ds.timestamp AS timestamp
+			FROM
+				ds
+			LEFT JOIN pair_stats_30m AS ps ON ps. "timestamp" < ds.timestamp
```

Comment:
Nit: adding the condition to this join can improve performance
```
 and ps."timestamp" >= ds.timestamp-%d
```
truncBy to %d

about `Execution Time: 1045.355 ms` to `Execution Time: 7.578 ms`

Reply:
but this condition prevents taking liquidities from before the specified interval..

Reply:
I think we cannot assume pairs always exist between 
`ps."timestamp" >= ds.timestamp-%d` and `ps. "timestamp" < ds.timestamp`

Reply:
> but this condition prevents taking liquidities from before the specified interval.
yes... it was the problem.
I am thinking a better way. Thank you for your suggestion!